File: Crio/crio_programming_interview_problems-master/ReplaceWords/ReplaceWords.py
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS

from collections import defaultdict
from functools import reduce

# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def replaceWords(words, sentence):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    Trie = lambda: defaultdict(Trie)
    trie = Trie()
    End = True
    for root in words:
        reduce(dict.__getitem__, root, trie)[End] = root

    def replace(word):
        cur = trie
        for letter in word:
            if letter not in cur or End in cur: 
                break
            cur = cur[letter]
        return cur.get(End, word)

    return " ".join(map(replace, sentence.split()))
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def main():
    n = int(input())
    words = input().strip().split()
    sentence  = input()
    ans  = replaceWords(words, sentence)
    print(ans)

if __name__ == "__main__":
    main()

File: Crio/crio_programming_interview_problems-master/ReplaceWords/ReplaceWords.java
import java.util.*;

public class ReplaceWords {
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    public static class Node{
        boolean end;
        Node[] child;
        Node(){
            child = new Node[26];
        }
    }
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    public static String replaceWords(ArrayList<String>dict, String sentence){
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        StringBuilder ans = new StringBuilder();
        Node head = new Node();
        for (String key : dict) {
            Node cur = head;
            for (char ch : key.toCharArray()) {
                if(cur.child[ch-'a']==null){
                    cur.child[ch-'a'] = new Node();
                }
                cur = cur.child[ch-'a'];
            }
            cur.end = true;
        }
        for (String word : sentence.split("\\s+")){
            Node cur = head;
            StringBuilder temp = new StringBuilder();
            for (char ch : word.toCharArray()) {
                if(cur.end){
                    break;
                }
                if(cur.child[ch-'a']!=null){
                    cur = cur.child[ch-'a'];
                }
                else{
                    temp.delete(0, temp.length());
                    temp.append(word);
                    break;
                }
                temp.append(ch);
            }
            temp.append(" ");
            ans.append(temp);
        }
        return ans.toString();
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        ArrayList<String>dict = new ArrayList<String>();
        for (int i = 0; i < n; i++) {
            dict.add(sc.next());
        }
        String sentence = sc.nextLine();
        sentence = sc.nextLine();
        String ans = replaceWords(dict,sentence);
        System.out.print(ans);
        sc.close();
    }    
}
